[SPARK-32463][SQL][DOCS] Add "Type Conversion" section in "Supported Data Types" of SQL docs

planga82 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Add new documentation about how spark treate data type compatibility

![image](https://user-images.githubusercontent.com/12819544/98147843-ab391400-1ec4-11eb-9bc9-bd929079ecc3.png)
![image](https://user-images.githubusercontent.com/12819544/98148062-b9873000-1ec4-11eb-9dc6-a70af8b46450.png)
![image](https://user-images.githubusercontent.com/12819544/98148229-c4da5b80-1ec4-11eb-96b8-423d5a3ed87f.png)
![image](https://user-images.githubusercontent.com/12819544/98148410-d1f74a80-1ec4-11eb-9640-50ab78187f45.png)
![image](https://user-images.githubusercontent.com/12819544/98148560-dcb1df80-1ec4-11eb-919b-0c25b1c39ad6.png)



### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
It's interesting for the final users

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Only documentation

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
Not needed

pablolanga-stratio committed:
[SPARK-32463][DOCS] SQL data type compatibility

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

planga82 commented:
@huaxingao Any comments are welcome. Thanks

HyukjinKwon commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,33 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+
+For example, if you have an add operation between an integer and a float, the integer will be treated as a float, the least common compatible type, resulting the operation in a float.
+
+The most common operations where this hierarchy is applied are:
```
I believe we have much more complicated rules than here ..  see `TypeCoercion.scala`.

planga82 replied:
Yes, I was trying to simplify and put in this list the most popular "operations" but I can extend it. I'm going to review it. Thanks!

planga82 replied:
I'm not sure if explain with a lot of detail this cases is useful. Possibly it's better to remove this list to avoid misunderstand. @HyukjinKwon what do you think?

huaxingao replied:
@gatorsmile WDYT? Do you want to have a table to list the most common rules like the one in https://www.ibm.com/support/knowledgecenter/SSEPGG_10.1.0/com.ibm.db2.luw.sql.ref.doc/doc/r0008477.html?

maropu replied:
Have you checked the other docs, e.g., Oracle, SQL server, ...? If they've already described  type coercsion, I think we are able to refer to them.

planga82 replied:
I have checked Oracle and SQL server documentation.
SQL server don't detail all implicit conversions:
https://docs.microsoft.com/en-us/sql/t-sql/data-types/data-type-conversion-database-engine?view=sql-server-ver15
https://docs.microsoft.com/en-us/sql/t-sql/data-types/data-type-precedence-transact-sql?view=sql-server-ver15
But Oracle have a more detailed rules (see "Implicit Data Conversion")
https://docs.oracle.com/cd/B28359_01/server.111/b28286/sql_elements002.htm#SQLRF00214

maropu replied:
Ah, I see. IMO the matrix format of the Oracle doc looks better.

huaxingao replied:
I guess maybe keep the current Hierarchy compatible types table because it shows the best to worst order, and also add a matrix format data conversion table like Oracle doc? IBM doc actually has two tables, one for Data Type Conversion Precedence List, the other is for casting between data types.

planga82 replied:
I think it's fine to have both tables. I'm going to update the PR to include it

pablolanga-stratio committed:
Add matrix compatible types

planga82 commented:
Updated with the matrix like Oracle Docs

maropu commented:
@planga82 Could you put the screenshot of the updated doc for reviews in the PR description?

maropu commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,49 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+
+#### Data type compatibility
+
+The following is the hierarchy of data type compatibility and the possible implicit conversions that can be made. In an operation involving different and compatible data types, these will be promoted to the lowest common top type to perform the operation.
```
The current description looks ambiguous and many topics get mixed up, I think. What's the topics you would like to describe in this section? If you want to describe type conversion with the default mode (ansi=false), I think we need to pick up three categories: [explicit casting, type coercion, and store assignment casting](https://spark.apache.org/docs/latest/sql-ref-ansi-compliance.html#type-conversion). Anyway, we need a clearer structure to describe type behaviours for easy-to-read user documents.

planga82 replied:
Thank you very much! I didn't know about Ansi property. I will redo the PR with this information and try to explain it with a better structure and also include a screenshot

pablolanga-stratio committed:
Refactor

planga82 commented:
I have done a refactor, I tried to explain more clear the concepts

![image](https://user-images.githubusercontent.com/12819544/95361866-2f4bac00-08c5-11eb-8375-efdab32e8440.png)
![image](https://user-images.githubusercontent.com/12819544/95361898-38d51400-08c5-11eb-94cd-5aa22509528e.png)

maropu commented:
Could you move the screenshots into the PR description?

maropu commented on docs/sql-ref-datatypes.md:
````diff
@@ -314,3 +314,51 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
 +---------+---------+
 ```
+
+#### Data type compatibility
````
`Data type compatibility` => `Type Conversion`?

maropu replied:
How about organizing the structure of this section by referring to the Oracle doc like this? https://docs.oracle.com/cd/B28359_01/server.111/b28286/sql_elements002.htm#SQLRF51043
```
#### Type Conversion
<What does "Type Conversion" means? How does Spark handle type conversion? brabrabra...>

#### Type Coercion
<Coercion Matrix> 

##### Type Coercion Examples
<examples...>

#### Explicit Casting and Store Assignment Casting
<Casting Matrix> 

##### Type Casting Examples
<examples...>
```
cc: @gatorsmile @HyukjinKwon @huaxingao

maropu commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,51 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+
+For example, if you have an add operation between an integer and a float, the integer will be treated as a float, the least common compatible type, resulting the operation in a float.
+
+|Data type|Hierarchy compatible types|
```
Could we use a matrix form for type coercion, too?

maropu commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,51 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+|DateType |*         |*         |X           |X        |X         |X          |X          |           |            |X             |--      |
+
+X: Conversion allowed (cast ByteType in ShortType)  
+*: An overflow can occur, check ANSI compliance for the result in this case (cast ShortType in ByteType)
```
The overflow behaviour is different between ANSI/non-ANSI modes. Could you explain about it here and add a link to the ANSI page? https://github.com/apache/spark/blob/master/docs/sql-ref-ansi-compliance.md#type-conversion We might need a subsection for the explanation.

pablolanga-stratio committed:
Re-structure sections

planga82 commented:
I have restructured the sections, I think it is more complete as you proposed. Any more ideas? Thank you!

![image](https://user-images.githubusercontent.com/12819544/95556788-c3279000-0a0b-11eb-8e3a-e75528c4d7b1.png)
![image](https://user-images.githubusercontent.com/12819544/95556827-d2a6d900-0a0b-11eb-84fa-5dd07c3042c3.png)
![image](https://user-images.githubusercontent.com/12819544/95556843-de929b00-0a0b-11eb-87c9-cfe1a03b82a2.png)
![image](https://user-images.githubusercontent.com/12819544/95556871-e7836c80-0a0b-11eb-9b0c-a81de0a744c4.png)

huaxingao commented:
This looks good to me overall. Thank for the good work! @planga82 
A couple of comments: 
1. Could you please put a `;` in the end of the SQL statements in the examples?
2. I quickly tried a few explicit casting, seems I can cast Boolean to Timestamp OK?

pablolanga-stratio committed:
Fix sql statement examples

planga82 commented:
Thanks @huaxingao !
> I quickly tried a few explicit casting, seems I can cast Boolean to Timestamp OK?
I noticed it, but I didn't put it in because even though it was allowed, I thought it didn't make sense. Shall we put it in anyway?

maropu commented on docs/sql-ref-datatypes.md:
````diff
@@ -314,3 +314,128 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
 +---------+---------+
 ```
+
+### Type conversion
````
`conversion` -> `Conversion`

maropu commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,128 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+
+### Type conversion
+
+In general, an expression can contain different data types, type conversion is the transformation of some data types into others in order to solve the expressions. 
```
`... can contain different data types, type conversion ...` -> `... can contain different data types and type conversion ...`?

maropu replied:
`in order to solve the expressions.` -> `in order to resolve type mismatches.`?

maropu commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,128 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+Spark supports both implicit conversions by type coercion and explicit conversions by explicit casting and store assignment casting.
+
+#### Type coercion in operations between different types 
+
```
Could you describe what' a type coercion at the beginning?

maropu commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,128 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+|**LongType**   |LongType   |LongType   |LongType    |--        |FloatType |DoubleType |DoubleType |
+|**FloatType**  |FloatType  |FloatType  |FloatType   |FloatType |--        |DoubleType |DoubleType |
+|**DoubleType** |DoubleType |DoubleType |DoubleType  |DoubleType|DoubleType|--         |DoubleType |
+|**StringType** |DoubleType |DoubleType |DoubleType  |DoubleType|DoubleType|DoubleType |--         |
```
This matrix means specific type coercion rules in binary arithmetic operations, right? Cold you add a matrix for type coercion in more general cases, too, like `Table 2-10 Implicit Type Conversion Matrix` in the Oracle doc? You could refer to the rules in `TypeCoercion.scala` https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L983-L1064

maropu commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,128 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+|**DoubleType** |DoubleType |DoubleType |DoubleType  |DoubleType|DoubleType|--         |DoubleType |
+|**StringType** |DoubleType |DoubleType |DoubleType  |DoubleType|DoubleType|DoubleType |--         |
+
+The case of DecimalType, is treated differently, for example, there is no common type for double and decimal because double's range is larger than decimal, and yet decimal is more precise than double, but in an expresion, we would cast the decimal into double.
```
Could you describe more about decimal type coercion by referring to `DecimalPrecision`?  https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala#L29-L63

planga82 replied:
Great information, I have included DecimalType in all matrix.

maropu commented on docs/sql-ref-datatypes.md:
````diff
@@ -314,3 +314,128 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+
+```
+
+#### Explicit casting and store assignment casting
````
cc: @gengliangwang

maropu replied:
`Explicit casting and store assignment casting` -> `Explicit Casting and Store Assignment Casting`

maropu commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,128 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+X: Conversion allowed (cast ByteType in ShortType)  
+*: An overflow can occur (cast ShortType in ByteType)
+
+If an overflow occurs and ANSI compliance is activated (spark.sql.ansi.enabled is set to true for casting or spark.sql.storeAssignmentPolicy=ANSI for store assignment casting) an exception will be thrown. 
```
spark.sql.ansi.enabled -> \`spark.sql.ansi.enabled\`

maropu replied:
spark.sql.storeAssignmentPolicy=ANSI -> \`spark.sql.storeAssignmentPolicy=ANSI\`

maropu commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,128 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
++---------------+---------+-------+
+
+SELECT MONTHS_BETWEEN(dateColumn,timestampColumn) FROM dateTable;
+
```
Could you put the output for following the other doc examples?

planga82 replied:
I don't added it because the result is an integer and it don't add anything interesting for the reader about type coercion between timestamp and date types. don't you think?

maropu commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,128 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+In general, an expression can contain different data types, type conversion is the transformation of some data types into others in order to solve the expressions. 
+Spark supports both implicit conversions by type coercion and explicit conversions by explicit casting and store assignment casting.
+
+#### Type coercion in operations between different types 
```
`Type coercion in operations between different types ` -> Type Coercion in Operations between Different Types

maropu commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,128 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+|**DateType**      |--           |TimestampType |
+|**TimestampType** |TimestampType|--            |
+
+#### Type coercion examples
```
`Type coercion examples` -> `Type Coercion Examples`

maropu commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,128 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+If an overflow occurs and ANSI compliance is activated (spark.sql.ansi.enabled is set to true for casting or spark.sql.storeAssignmentPolicy=ANSI for store assignment casting) an exception will be thrown. 
+Otherwise, a truncate value will be used. See more on [Ansi Compliance](sql-ref-ansi-compliance.html#type-conversion).
+
+#### Type Casting examples
```
`Type Casting examples` -> `Type Casting Examples`

maropu commented on docs/sql-ref-datatypes.md:
````diff
@@ -314,3 +314,128 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+
+```
+
+#### Store assignment casting examples
````
`Store assignment casting examples` -> `Store Assignment Casting Examples`

pablolanga-stratio committed:
Refactor with PR comments

planga82 commented:
I have update the PR with all comments. very useful, thanks!

pablolanga-stratio committed:
Fix end of line

huaxingao commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,151 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+Type Coercion refers to the automatic or implicit conversion of values from one type to another when you need to to resolve type mismatches.
+The following matrix shows the resulting type to which they are implicitly converted to resolve an expression involving different data types.
+
+**Numeric expressions**:
```
super nit: Numeric expressions -> Numeric Expressions?

huaxingao commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,151 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+**Note 1**: DecimalType(precision,scale)   
+**Note 2**: In these cases DecimalType can lose precision, there is no common type for decimal and double because double's range is larger than decimal, and yet decimal is more precise than double so when we cast Decimaltype into DobleType it could lose precision.
+
+**Time expressions**:
```
super nit: Time expressions -> Time Expressions?

huaxingao commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,151 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+|**TimestampType** |TimestampType|--            |
+
+
+**Possible implicit conversions**:
```
super nit: Possible implicit conversions -> Possible Implicit Conversions?

huaxingao commented:
@gatorsmile Could you please take a quick look?

gatorsmile commented:
cc @gengliangwang @cloud-fan who is working on this area recently.

pablolanga-stratio committed:
Fix capital letters

gengliangwang commented:
![image](https://user-images.githubusercontent.com/12819544/95687803-db5f0100-0bfd-11eb-952a-c714f880556d.png)
This is not true. The type conversion rules are more complex than that.

```
spark-sql> explain select 1 in (2, 'a');
*(1) Project [false AS (CAST(1 AS STRING) IN (CAST(2 AS STRING), CAST(a AS STRING)))#19]

spark-sql> explain select 1 = '2';
*(1) Project [false AS (1 = CAST(2 AS INT))#11]

spark-sql> explain select 1 + '2';
*(1) Project [3.0 AS (CAST(1 AS DOUBLE) + CAST(2 AS DOUBLE))#17]
```

gengliangwang reviewed (changes requested):
The type conversion rules are quite complex. The current table for numeric/string conversion is not right.

planga82 commented:
> This is **not true**. The type conversion rules are more complex than that.
> 
> ```
> spark-sql> explain select 1 in (2, 'a');
> *(1) Project [false AS (CAST(1 AS STRING) IN (CAST(2 AS STRING), CAST(a AS STRING)))#19]
> 
> spark-sql> explain select 1 = '2';
> *(1) Project [false AS (1 = CAST(2 AS INT))#11]
> 
> spark-sql> explain select 1 + '2';
> *(1) Project [3.0 AS (CAST(1 AS DOUBLE) + CAST(2 AS DOUBLE))#17]
> ```

Ok, I see, do you think it's better to drop StringType from this matrix? Do we have other differences in other types?

gengliangwang commented:
> Ok, I see, do you think it's better to drop StringType from this matrix? Do we have other differences in other types?

Yes, but still we need to describe the behavior of type conversion between string type and numeric/date/timestamp type.

pablolanga-stratio committed:
Explain StringType behavior

planga82 commented:
I have updated the document with the StringType behavior and also added new examples.

![image](https://user-images.githubusercontent.com/12819544/96506836-34f3ab00-1250-11eb-9e52-f03dacab29b0.png)
![image](https://user-images.githubusercontent.com/12819544/96506863-45a42100-1250-11eb-9265-94b962ea153c.png)
![image](https://user-images.githubusercontent.com/12819544/96506896-52287980-1250-11eb-95ac-5b139698d067.png)
![image](https://user-images.githubusercontent.com/12819544/96506952-5fddff00-1250-11eb-80d5-2d19ee362b83.png)
![image](https://user-images.githubusercontent.com/12819544/96506976-69fffd80-1250-11eb-8191-25ecc2c1cde4.png)

planga82 commented:
@gengliangwang Do you think the explanation is correct now? Thanks

gengliangwang commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,200 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+
+    |               |ByteType   |ShortType  |IntegerType |LongType   |FloatType    |DoubleType  |
+    |---------------|-----------|-----------|------------|-----------|-------------|------------|
+    |**StringType** |DoubleType |DoubleType |DoubleType  |DoubleType |DoubleType   |DoubleType  |
```
I think we need to make it clear that the String type will be implicit cast as Double type.

gengliangwang commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,200 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+
+    |               |ByteType   |ShortType  |IntegerType |LongType   |FloatType    |DoubleType  |DecimalType |DateType             |TimestampType             |
+    |---------------|-----------|-----------|------------|-----------|-------------|------------|------------|---------------------|--------------------------|
+    |**StringType** |ByteType   |ShortType  |IntegerType |LongType   |FloatType    |DoubleType  |DoubleType  |DateType<sup>1</sup> |TimestampType<sup>1</sup> |
```
ditto

gengliangwang commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,200 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+
+    **Note 1**: If `spark.sql.legacy.typeCoercion.datetimeToString` is true, DateType and TimestampType will be casted to StringType
+    
+* IN Expressions: Expressions like `x IN list_values`.  If the list of values has a StringType element, all the elements will be casted to StringType  
```
I think we need to write down all the expressions here. 
```
In
Except
Intersect
Union
CreateArray
Concat
Sequence
MapConcat
CreateMap
CaseWhen
If
```
Please check if there is any missing one

planga82 replied:
I have included the expressions where I have checked that type casting was done in StringType. 
I have left out the following:
* greatest, array_intersect, array_union, array_except: Don't support diffent types, so there will be no casting.
* Sequence: Don't support StringType

pablolanga-stratio committed:
Update functions in StringType Behavior

planga82 commented:
@gengliangwang I have introduced descriptions to clarify and listed the expressions

![image](https://user-images.githubusercontent.com/12819544/98040246-da457c00-1e17-11eb-8875-8dc1a466c204.png)
![image](https://user-images.githubusercontent.com/12819544/98040320-f8ab7780-1e17-11eb-8bce-bd3fd51ac140.png)
![image](https://user-images.githubusercontent.com/12819544/98040346-082ac080-1e18-11eb-8632-5ab8c4855b45.png)
![image](https://user-images.githubusercontent.com/12819544/98040369-11b42880-1e18-11eb-883d-9c581372f451.png)
![image](https://user-images.githubusercontent.com/12819544/98040508-4a540200-1e18-11eb-883d-cf2bc120016d.png)
![image](https://user-images.githubusercontent.com/12819544/98040534-52ac3d00-1e18-11eb-951d-eec1013c729f.png)

cloud-fan commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,206 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+
+### Type Conversion
+
+In general, an expression can contain different data types and type conversion is the transformation of some data types into others in order to resolve type mismatches. 
```
```
Type conversion turns the values of one data type to another data type. Spark needs to perform
type conversions if users explicitly ask to do so via the CAST operator, or to resolve data type
mismatch in operators, functions, and table writing implicitly.
```

cloud-fan commented on docs/sql-ref-datatypes.md:
```diff
@@ -314,3 +314,206 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
+
+
+
+#### Explicit Casting and Store Assignment Casting
```
Ansi mode is off by default, and Store Assignment is using ANSI policy by default. That said, these two have different behaviors by default.

planga82 replied:
Ok, thanks, I'll explain it in the table notes

pablolanga-stratio committed:
Update descriptions

planga82 commented:
Image updated in the description of the PR

gatorsmile commented:
cc @gengliangwang Could you take a look? We should add it to our doc.
